support/agent/sim.go

```diff
@@ -0,0 +1,251 @@
+	if amt.LessThanEqual(big.Zero()) {
+		return nil
+	}
+
+	rewardParams := builtin.ApplyRewardParams{
+		Reward:  amt,
+		Penalty: big.Zero(),
+	}
+	_, code := s.v.ApplyMessage(builtin.RewardActorAddr, addr, amt, builtin.MethodsMiner.ApplyRewards, &rewardParams)
+	if code != exitcode.Ok {
+		return errors.Errorf("exitcode %d: reward message failed:\n%s\n", code, strings.Join(s.v.GetLogs(), "\n"))
+	}
+	return nil
+}
+
+func ComputePowerTable(v *vm.VM, miners []*MinerAgent) (PowerTable, error) {
+	pt := PowerTable{}
+
+	var rwst reward.State
+	if err := v.GetState(builtin.RewardActorAddr, &rwst); err != nil {
+		return PowerTable{}, err
+	}
+	pt.blockReward = rwst.ThisEpochReward
+
+	var st power.State
+	if err := v.GetState(builtin.StoragePowerActorAddr, &st); err != nil {
+		return PowerTable{}, err
+	}
+	pt.totalQAPower = st.TotalQualityAdjPower
+
+	for _, miner := range miners {
+		if claim, found, err := st.GetClaim(v.Store(), miner.IDAddress); err != nil {
+			return pt, err
+		} else if found {
+			if sufficient, err := st.MinerNominalPowerMeetsConsensusMinimum(v.Store(), miner.IDAddress); err != nil {
+				return pt, err
+			} else if sufficient {
+				pt.minerPower = append(pt.minerPower, MinerPowerTable{miner.IDAddress, claim.QualityAdjPower})
+			}
+		}
+	}
+	return pt, nil
+}
+
+func (s *Sim) WinCount(minerPower abi.StoragePower, totalPower abi.StoragePower) uint64 {
+	E := big2.NewRat(5, 1)
+	lambdaR := new(big2.Rat)
+	lambdaR.SetFrac(minerPower.Int, totalPower.Int)
+	lambdaR.Mul(lambdaR, E)
+	lambda, _ := lambdaR.Float64()
+
+	h := s.rnd.Float64()
+	rhs := 1 - poissonPMF(lambda, 0)
+
+	winCount := uint64(0)
+	for rhs > h {
+		winCount++
+		rhs -= poissonPMF(lambda, winCount)
+	}
+	return winCount
+}
+
+func poissonPMF(lambda float64, k uint64) float64 {
+	fk := float64(k)
+	return (math.Exp(-lambda) * math.Pow(lambda, fk)) / fact(fk)
+}
+
+func fact(k float64) float64 {
+	fact := 1.0
+	for i := 2.0; i <= k; i += 1.0 {
+		fact *= i
+	}
+	return fact
+}
+
+func (s *Sim) GetVM() *vm.VM {
+	return s.v
+}
+
+func (s *Sim) createMiner(owner address.Address, cfg MinerAgentConfig) Message {
+	return Message{
+		From:   owner,
+		To:     builtin.StoragePowerActorAddr,
+		Value:  s.Config.AccountInitialBalance, // miner gets all account funds
+		Method: builtin.MethodsPower.CreateMiner,
+		Params: &power.CreateMinerParams{
+			Owner:         owner,
+			Worker:        owner,
+			SealProofType: cfg.ProofType,
+		},
+		ReturnHandler: func(_ *vm.VM, msg Message, ret cbor.Marshaler) error {
+			createMinerRet, ok := ret.(*power.CreateMinerReturn)
+			if !ok {
+				return errors.Errorf("create miner return has wrong type: %v", ret)
+			}
+
+			params := msg.Params.(*power.CreateMinerParams)
+			if !ok {
+				return errors.Errorf("create miner params has wrong type: %v", msg.Params)
+			}
+
+			miner := NewMinerAgent(params.Owner, params.Worker, createMinerRet.IDAddress, createMinerRet.RobustAddress, s.rnd, cfg)
```

Comment:
Ok, this use of a return handler seems more legitimate. I would love to find a way to avoid it, inspecting state instead, but it justifies the mechanism for now.

As noted above, I think this could be pulled out into a miner factor agent.